lib/src/native/realm_core.dart

```diff
@@ -480,15 +496,56 @@ class _RealmCore {
 
   static void _syncErrorHandlerCallback(Object userdata, Pointer<realm_sync_session> session, realm_sync_error error) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
+    final syncError = error.toSyncError(syncConfig);
 
-    final syncError = error.toSyncError();
+    if (syncError is ClientResetError) {
+      syncConfig.clientResetHandler.onManualReset?.call(syncError);
+    } else {
+      syncConfig.syncErrorHandler(syncError);
+    }
+  }
 
-    if (syncError is SyncClientResetError) {
-      syncConfig.syncClientResetErrorHandler.callback(syncError);
+  static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
+    final syncConfig = userdata as FlexibleSyncConfiguration;
+    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset;
+    if (beforeResetCallback == null) {
+      _invokeNativeFunction(unlockCallbackFunc);
       return;
     }
+    final realm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(realmHandle));
+    _continueWhenComplete(() => beforeResetCallback(realm), unlockCallbackFunc);
+  }
 
-    syncConfig.syncErrorHandler(syncError);
+  static void _syncAfterResetCallback(Object userdata, Pointer<shared_realm> beforeHandle, Pointer<realm_thread_safe_reference> afterReference, bool didRecover,
+      Pointer<Void> unlockCallbackFunc) {
+    final syncConfig = userdata as FlexibleSyncConfiguration;
+    final afterResetCallback = didRecover ? syncConfig.clientResetHandler.onAfterRecovery : syncConfig.clientResetHandler.onAfterDiscard;
+    if (afterResetCallback == null) {
+      _invokeNativeFunction(unlockCallbackFunc);
+      return;
+    }
+    final beforeRealm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(beforeHandle));
+    final afterRealm = RealmInternal.getUnowned(
+      syncConfig,
+      RealmHandle._unowned(_realmLib.realm_from_thread_safe_reference(afterReference, scheduler.handle._pointer)),
+    );
+    _continueWhenComplete(() => afterResetCallback(beforeRealm, afterRealm), unlockCallbackFunc);
+  }
+
+  static void _invokeNativeFunction(Pointer<Void> nativeFunction, [Object? error]) {
+    if (error != null) {
+      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
```

Comment:
when are we calling realm_get_last_error to retrieve this error. Aren't we leaking this handle or aren't we setting last error incorrectly for the next Core call to fail

Reply:
This error is not provided when we are getting the last error. It is only logged that 'User-provided callback failed'.